src/core/validate-graph.ts

```diff
@@ -30,4 +32,30 @@ export function validateGraph(
     (pkgId) => !pkgNodes[pkgId] || pkgNodes[pkgId].size === 0,
   );
   assert(pkgsWithoutInstances.length === 0, 'not all pkgs have instance nodes');
+
+  for (const pkgId in pkgs) {
+    try {
+      validatePackageURL(pkgs[pkgId] as types.PkgInfo);
+    } catch (e) {
+      throw new ValidationError(`invalid pkg ${pkgId}: ${e}`);
+    }
+  }
+}
+
+export function validatePackageURL(pkg: types.PkgInfo): void {
+  if (!pkg.purl) {
+    return;
+  }
+
+  try {
+    const purlPkg = PackageURL.fromString(pkg.purl);
+
+    assert(pkg.name === purlPkg.name, `name and packageURL name do not match`);
```

Comment:
Ah right, forgot about that! 
Should I add a special case for the maven type or should I exclude the assertion completely? I would vote for the first option :)

Reply:
Maven special case (with tests) :D

Reply:
Are you sure `com.namespace:foo` is the right format?
If so, this would actually be fine, because purl uses a `/` to separate namespace and name:
```
pkg:maven/org.apache.xmlgraphics/batik-anim@1.9.1?packaging=sources
```

And if I do `PackageURL.fromString("pkg:maven/com.namespace:foo")` I get a purl with the name `com.namespace:foo`

Reply:
> Are you sure `com.namespace:foo` is the right format?

It's not right, it's just the Snyk format. There's a mismatch here.

> And if I do `PackageURL.fromString("pkg:maven/com.namespace:foo")` I get a purl with the name `com.namespace:foo`

Yeh, and that's wrong, no one should be inputting `"pkg:maven/com.namespace:foo"`. We're munging the purl namespace (maven group id) into the name. A valid `pkgInfo` should be:

```
{
  name: 'com.example:foo',
  version: '1.0.0',
  purl: 'pkg:maven/com.example/foo@1.0.0'
}
```

Reply:
i.e. Snyk:
```
{ name: 'com.example:foo', version: '1.0.0' }
```
Purl:
```
{ name: 'foo', namespace: 'com.example', version: '1.0.0' }
```

Reply:
https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#maven

Reply:
Sorry, there's a slight misunderstanding here, I might have not explained myself properly.

> It's not right, it's just the Snyk format. There's a mismatch here.

Yes, that's what I meant. 

> Yeh, and that's wrong, no one should be inputting "pkg:maven/com.namespace:foo". We're munging the purl namespace (maven group id) into the name. A valid pkgInfo should be:

Ah, I think now i get it. We're still expecting the purl to be correct (e.g. `pkg:maven/com.example/foo@1.0.0`), but the name will be `com.example:foo`. Makes sense - that's the only case where we'd need a separate logic to verify it 😄

Reply:
> that's the only case where we'd need a separate logic to verify it 😄

Right now, yes, maybe, probably not... We've almost certainly done other silly things in this space that will need to be dealt with over time as people adopt purl in their graphs. i.e. we've encoded too many things into `name` because that's all that the dep-graph had.